packages/astro/src/actions/index.ts

```diff
@@ -0,0 +1,87 @@
+import { existsSync } from 'node:fs';
+import { mkdir, readFile, writeFile } from 'node:fs/promises';
+import type { AstroIntegration } from '../@types/astro.js';
+import { ACTIONS_TYPES_FILE, RESOLVED_VIRTUAL_MODULE_ID, VIRTUAL_MODULE_ID } from './consts.js';
+import type { Plugin as VitePlugin } from 'vite';
+
+const name = 'astro:actions';
+
+export default function astroActions(): AstroIntegration {
+	return {
+		name,
+		hooks: {
+			async 'astro:config:setup'(params) {
+				const stringifiedActionsPath = JSON.stringify(
+					new URL('actions', params.config.srcDir).pathname
+				);
+				params.updateConfig({
+					vite: {
+						define: {
+							'import.meta.env.ACTIONS_PATH': stringifiedActionsPath,
+						},
+						plugins: [vitePluginActions],
+					},
+				});
+
+				params.injectRoute({
+					pattern: '/_actions/[...path]',
+					entrypoint: 'astro/actions/runtime/route.js',
+					prerender: false,
+				});
+
+				params.addMiddleware({
+					entrypoint: 'astro/actions/runtime/middleware.js',
+					order: 'pre',
+				});
+
+				await typegen({
+					stringifiedActionsPath,
+					root: params.config.root,
+				});
+			},
+		},
+	};
+}
+
+const vitePluginActions: VitePlugin = {
+	name: 'astro:actions',
+	enforce: 'pre',
+	resolveId(id) {
+		if (id === VIRTUAL_MODULE_ID) {
+			return RESOLVED_VIRTUAL_MODULE_ID;
+		}
+	},
+	async load(id, opts) {
+		if (id !== RESOLVED_VIRTUAL_MODULE_ID) return;
+
+		let code = await readFile(
+			new URL('../../actions-module.template.mjs', import.meta.url),
+			'utf-8'
+		);
+		if (opts?.ssr) {
+			code += `\nexport * from 'astro/actions/runtime/virtual/server.js';`;
```

Comment:
Genuine question: this is a barrel file. Is it safe in terms of performance to use a barrel file inside a virtual module?

Reply:
Hm, I hadn't considered performance for `*` exports. Enumerating each export individually would mean another place to update exports, which could cause future bugs. How would we improve this?